src/java_bytecode/java_local_variable_table.cpp

```diff
@@ -261,45 +276,76 @@ static void populate_predecessor_map(
+    // they cannot have live predecessor ranges: they are initialized by the JVM
+    // and no other live variable range can flow into them.
+    if(it->is_parameter)
       continue;
 
+    msg.debug() << "ppm: processing var idx " << it->var.index
+                << " name '" << it->var.name << "' start-pc "
+                << it->var.start_pc << " len " << it->var.length
+                << "; holes " << it->holes.size() << messaget::eom;
+
     // Find the last instruction within the live range:
     unsigned end_pc=it->var.start_pc+it->var.length;
     auto amapit=amap.find(end_pc);
-    assert(amapit!=amap.begin());
+    INVARIANT(
+      amapit!=amap.begin(),
+      "current bytecode shall not be the first");
     auto old_amapit=amapit;
     --amapit;
     if(old_amapit==amap.end())
     {
-      assert(
-        end_pc>amapit->first &&
+      INVARIANT(
+        end_pc>amapit->first,
         "Instruction live range doesn't align to instruction boundary?");
     }
 
-    // Find vartable entries that flow into this one:
+    // Find vartable entries that flow into this one. For unknown reasons this
+    // loop iterates backwards, walking back from the last bytecode in the live
+    // range of variable it to the first one. For each value of the iterator
+    // "amapit" we search for instructions that jump into amapit's address
+    // (predecessors)
     unsigned new_start_pc=it->var.start_pc;
     for(; amapit->first>=it->var.start_pc; --amapit)
     {
       for(auto pred : amapit->second.predecessors)
       {
+        // pred is the address (byecode offset) of a instruction that jumps into
+        // amapit. Compute now a pointer to the variable-with-holes whose index
+        // equals that of amapit and which was alive on instruction pred, or a
+        // null pointer if no such variable exists (e.g., because no live range
+        // covers that instruction)
         auto pred_var=
           (pred<live_variable_at_address.size() ?
            live_variable_at_address[pred] :
-           0);
+           nullptr);
+
+        // Three cases are now possible:
+        // 1. The predecessor instruction is in the same live range: nothing to
+        // do.
         if(pred_var==&*it)
         {
-          // Flow from within same live range?
           continue;
         }
+        // 2. The predecessor instruction is in no live range among those for
+        // variable slot it->var.index
         else if(!pred_var)
         {
-          // Flow from out of range?
-          // Check if this is an initialiser, and if so expand the live range
+          // Check if this is an initializer, and if so expand the live range
           // to include it, but don't check its predecessors:
           auto inst_before_this=amapit;
-          assert(inst_before_this!=amap.begin());
+          INVARIANT(
+            inst_before_this!=amap.begin(),
+            "we shall not be on the first bytecode of the method");
```

Comment:
Suggestion (should probably go into coding guideline): Can we consistently start these messages with an upper case letter?